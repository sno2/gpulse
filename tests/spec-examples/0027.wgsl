// Declare a variable in the private address space, for storing an f32 value.
var<private> x: f32;

fn f() {
    // Declare a variable in the function address space, for storing an i32 value.
    var y: i32;

    // The name 'x' resolves to the module-scope variable 'x',
    // and has reference type ref<private,f32,read_write>.
    // Applying the unary '&' operator converts the reference to a pointer.
    // The access mode is the same as the access mode of the original variable, so
    // the fully specified type is ptr<private,f32,read_write>.  But read_write
    // is the default access mode for function address space, so read_write does not
    // have to be spelled in this case
    let x_ptr: ptr<private,f32> = &x;

    // The name 'y' resolves to the function-scope variable 'y',
    // and has reference type ref<private,i32,read_write>.
    // Applying the unary '&' operator converts the reference to a pointer.
    // The access mode defaults to 'read_write'.
    let y_ptr: ptr<function,i32> = &y;

    // A new variable, distinct from the variable declared at module scope.
    var x: u32;

    // Here, the name 'x' resolves to the function-scope variable 'x' declared in
    // the previous statement, and has type ref<function,u32,read_write>.
    // Applying the unary '&' operator converts the reference to a pointer.
    // The access mode defaults to 'read_write'.
    let inner_x_ptr: ptr<function,u32> = &x;
}
