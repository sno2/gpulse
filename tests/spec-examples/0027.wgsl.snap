ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 91, .end = 92 }, .access_mode = ast.AccessMode.read, .addr_space = ast.AddrSpace.private, .typ = ast.Node{ .identifier = ast.Span{ .start = 94, .end = 97 } }, .value = null } }
ast.Node{ .fn_decl = ast.FnDecl{ .name = ast.Span{ .start = 103, .end = 104 }, .params = {  }, .ret = null, .scope = { ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 200, .end = 201 }, .access_mode = ast.AccessMode.read, .addr_space = null, .typ = ast.Node{ .identifier = ast.Span{ .start = 203, .end = 206 } }, .value = null } }, ast.Node{ .let_decl = ast.LetDecl{ .name = ast.Span{ .start = 702, .end = 707 }, .typ = ast.Node{ .template = ast.Template{ .name = ast.Span{ .start = 709, .end = 712 }, .args = { ast.Node{ .identifier = ast.Span{ .start = 713, .end = 720 } }, ast.Node{ .identifier = ast.Span{ .start = 721, .end = 724 } } } } }, .value = ast.Node{ .ref = ast.UnaryOp{ .value = ast.Node{ .identifier = ast.Span{ .start = 729, .end = 730 } } } } } }, ast.Node{ .let_decl = ast.LetDecl{ .name = ast.Span{ .start = 990, .end = 995 }, .typ = ast.Node{ .template = ast.Template{ .name = ast.Span{ .start = 997, .end = 1000 }, .args = { ast.Node{ .identifier = ast.Span{ .start = 1001, .end = 1009 } }, ast.Node{ .identifier = ast.Span{ .start = 1010, .end = 1013 } } } } }, .value = ast.Node{ .ref = ast.UnaryOp{ .value = ast.Node{ .identifier = ast.Span{ .start = 1018, .end = 1019 } } } } } }, ast.Node{ .var_decl = ast.VarDecl{ .name = ast.Span{ .start = 1106, .end = 1107 }, .access_mode = ast.AccessMode.read, .addr_space = null, .typ = ast.Node{ .identifier = ast.Span{ .start = 1109, .end = 1112 } }, .value = null } }, ast.Node{ .let_decl = ast.LetDecl{ .name = ast.Span{ .start = 1404, .end = 1415 }, .typ = ast.Node{ .template = ast.Template{ .name = ast.Span{ .start = 1417, .end = 1420 }, .args = { ast.Node{ .identifier = ast.Span{ .start = 1421, .end = 1429 } }, ast.Node{ .identifier = ast.Span{ .start = 1430, .end = 1433 } } } } }, .value = ast.Node{ .ref = ast.UnaryOp{ .value = ast.Node{ .identifier = ast.Span{ .start = 1438, .end = 1439 } } } } } } } } }
